.travis.yml

```diff
@@ -34,9 +34,43 @@ install:
 jobs:
   include:
     - if: type = pull_request
-      name: checkpatch
+      # These two are combined because they both need a merge base and
+      # locally unshallowing .git/ seems time consuming for some reason.
+      # Without the merge base, the --xxxstats will either fail or be
+      # wrong and checkpatch prints confusing garbage.
+      name: checkpatch and age of git base
       script:
+        # Start with some visual and plain English context. A picture is
+        # worth thousand words and makes all the rest less abstract.
         - *short_log
+        - git --no-pager log --oneline --graph --decorate
+                                    --max-count=5 "${TRAVIS_BRANCH}"
+        # If the merge base is missing then the shallow depth
+        # optimization has _already_ failed because Travis' already
+        # fetched practically the entire history with a "git fetch
+        # github pulls/1234/merge" command. The very long duration of
+        # that earlier fetch merge command is hidden in Travis logs
+        # under the git clone command _wrongly_ advertised as quick!
+        - git merge-base "${TRAVIS_PULL_REQUEST_SHA}" "${TRAVIS_BRANCH}" ||
+              git fetch --unshallow https://github.com/"${TRAVIS_REPO_SLUG}"
+                "${TRAVIS_BRANCH}"
+        # Show how "late" the PR base is
+        - |
+          # Tests are run on the PR merged with this extra code from the
+          # (moving) target branch:
+          git diff --shortstat --dirstat \
+              "${TRAVIS_PULL_REQUEST_SHA}"..."${TRAVIS_BRANCH}"
+        # Use set -x to expand all values in the log.
+        # No double dot '..' so rev-list fails if a TRAVIS_ variable is missing
+        - set -x; behind=$(git rev-list --no-merges
+            ^"${TRAVIS_PULL_REQUEST_SHA}" "${TRAVIS_BRANCH}" | wc -l) ; set +x
```

Comment:
shouldnt this be to the result of merge base?

Reply:
I'm not parsing your comment sorry.

Just guessing: are you saying the merge base could be used here instead of TRAVIS_PULL_REQUEST_SHA? When there is only one merge-base then yes it wouldn't make any difference. If there are multiple merge bases then they would have to be excluded all, otherwise the `$behind` number would be higher. https://medium.com/@kevinwkds/git-recursive-merge-3fb14256768a has a good, simplified example.

Reply:
Ah yes, I forgot about the case where the parent is merged into the child and creates a second base

Reply:
Criss-cross a.k.a. "backmerges" can create very interesting situations. Here's one where old code gets (silently!) resurrected: https://bramcohen.livejournal.com/74462.html

There seems to be frequent backmerges in this branch.